fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/PruneComplexTypeUtil.java

```diff
@@ -0,0 +1,262 @@
+import com.starrocks.sql.optimizer.operator.scalar.CallOperator;
+import com.starrocks.sql.optimizer.operator.scalar.CollectionElementOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperatorVisitor;
+import com.starrocks.sql.optimizer.operator.scalar.SubfieldOperator;
+
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+public class PruneComplexTypeUtil {
+    // For example, we have a column col: MAP<INT, STRUCT<a INT, b STRUCT<c INT, d INT>>>
+    // And we have a sql: SELECT map_values(col).b.d, map_keys(col) from TABLE;
+    // Column col has a specific ColumnRefOperator
+    // map_values(col).b.d has three access paths:
+    // --- Access path 1: ComplexTypeAccessPath(MAP_VALUES)
+    // --- Access path 2: ComplexTypeAccessPath(STRUCT_SUBFIELD, "b")
+    // --- Access path 3: ComplexTypeAccessPath(STRUCT_SUBFIELD, "d")
+    // So it's access paths is: [Access path 1, Access path 2, Access path 3]
+    // map_keys(col) has one access paths:
+    // --- Access path 4: ComplexTypeAccessPath(MAP_KEYS)
+    // So it's access paths is: [Access path 4]
+    // Summary: ColumnRefOperator(col) has two access paths, we use access group to represent it.
+    // ColumnRefOperator(col):
+    // --- Access group: [[Access path 1, Access path 2, Access path 3], [Access path 4]]
+    protected static class Context {
+        // The same ColumnRefOperator may have multiple access paths for complex type
+        private final Map<ColumnRefOperator, List<ImmutableList<ComplexTypeAccessPath>>> accessGroups;
+
+        private boolean enablePruneComplexTypes;
+
+        public Context() {
+            this.accessGroups = new HashMap<>();
+            this.enablePruneComplexTypes = true;
+        }
+
+        public void setEnablePruneComplexTypes(boolean enablePruneComplexTypes) {
+            this.enablePruneComplexTypes = enablePruneComplexTypes;
+        }
+
+        public boolean getEnablePruneComplexTypes() {
+            return this.enablePruneComplexTypes;
+        }
+
+        public void addAccessPaths(ColumnRefOperator columnRefOperator,
+                                   ImmutableList<ComplexTypeAccessPath> accessPaths) {
+            accessGroups.putIfAbsent(columnRefOperator, new ArrayList<>());
+            accessGroups.get(columnRefOperator).add(accessPaths);
+        }
+
+        // Merge upper PhysicalNode's Context
+        public void addUpperContext(Context upperContext) {
+            setEnablePruneComplexTypes(upperContext.getEnablePruneComplexTypes());
+            if (!enablePruneComplexTypes) {
+                return;
+            }
+            Map<ColumnRefOperator, List<ImmutableList<ComplexTypeAccessPath>>> upperGroups = upperContext.getAccessGroups();
```

Comment:
getContext alread merge the AccessGroup, why do it again?

Reply:
This function only used by globalContext, used to merge all operator's context into a globalContext. So that we can prune ColumnRefOperator by global unique id.